projects/github-actions/test-results-to-slack/src/slack.js

```diff
@@ -156,6 +156,70 @@ async function getMessage( client, channelId, identifier ) {
 	return message;
 }
 
+/**
+ * Uploads a file to Slack using their three-step upload process.
+ * https://api.slack.com/messaging/files#uploading_files
+ *
+ * @param {object} client     - Slack client instance
+ * @param {string} filePath   - Path to the file to upload
+ * @param {string} channel_id - Channel ID where the file will be shared
+ * @param {string} thread_ts  - Thread timestamp for threading the file upload
+ * @return {Promise<object>} The response from the Slack API
+ */
+async function uploadFileToSlack( client, filePath, channel_id, thread_ts ) {
+	try {
+		// Get file stats for size and name
+		const stats = fs.statSync( filePath );
+		const filename = path.basename( filePath );
+
+		// Step 1: Get upload URL
+		const { ok, upload_url, file_id } = await client.files.getUploadURLExternal( {
+			filename,
+			length: stats.size,
+		} );
+		if ( ! ok ) {
+			throw new Error( 'Failed to get upload URL' );
+		}
+
+		// Step 2: Upload file to the URL
+		const fileStream = fs.createReadStream( filePath );
+		const fileBuffer = await new Promise( ( resolve, reject ) => {
+			const chunks = [];
+			fileStream.on( 'data', chunk => chunks.push( chunk ) );
+			fileStream.on( 'end', () => resolve( Buffer.concat( chunks ) ) );
+			fileStream.on( 'error', reject );
+		} );
+
+		const uploadResponse = await fetch( upload_url, {
+			method: 'POST',
+			body: fileBuffer,
+			headers: {
+				'Content-Type': 'application/octet-stream',
+			},
+		} );
```

Comment:
Looks like, instead of reading the stream into memory, it may be possible to do like this
```suggestion
		const fileStream = fs.createReadStream( filePath );
		const uploadResponse = await fetch( upload_url, {
			method: 'POST',
			body: fileStream,
			duplex: 'half',
			headers: {
				'Content-Type': 'application/octet-stream',
			},
		} );
```
Or, if that doesn't work or you'd rather have a Buffer anyway, you should be able to read it directly like
```suggestion
		const fileBuffer = await fs.promises.readFile( filePath );
		const uploadResponse = await fetch( upload_url, {
			method: 'POST',
			body: fileBuffer,
			headers: {
				'Content-Type': 'application/octet-stream',
			},
		} );
```